### PR TITLE
Enclose event read loop in try/finally to ensure close is done

### LINKED ISSUE #51: Properly handle close triggered by ping timeout
For some reason, the semantics of how close is handled for ping timeouts differs from other closes handling, including when triggered by server termination.  In the latter case, the implicit `read` loop exits and our termination handler runs.  In the former, the read throws an exception which bypasses our termination handling.

### PR DESCRIPTION
**Pull request by sfitts**
This ensures that we handle both close cases that exit the loop normally and those (like ping timeout) that result in an exception.

Fixes #51

### CODE REVIEW

File: src/main/python/vantiqsdk.py [hunk 1]
```diff
@@ -1392,73 +1392,74 @@ async def connect(self, do_pings: bool = True):
...
-                                                 resp['body'][0]['code'], resp['body'][0]['message'])
-                                ve = VantiqException('io.vantiq.python.connect.failed',
-                                                     'Connect call failed: {0} :: {1}:{2}',
-                                                     [resp['status'], resp['body'][0]['code'],
-                                                      resp['body'][0]['message']])
-                                self.connected_future.set_exception(ve)
-                                raise ve
-                            else:
+                try:
+                    if not self.connected:
```
Comment (sfitts) [inline src/main/python/vantiqsdk.py:1396]: Most of this is indentation

File: src/main/python/vantiqsdk.py [hunk 2]
```diff
@@ -1392,73 +1392,74 @@ async def connect(self, do_pings: bool = True):
...
-                        elif resp['status'] == 100:
-                            if request_id and request_id in self.callbacks.keys():
-                                self._vlog.debug('Message received via subscription.')
-                                callback = self.callbacks[request_id]
-                                await callback(self.MESSAGE, resp)
-
-            # Once we get here, any transient subscriptions will be gone, so we should reset our side as well
-            await self.unsubscribe_all()
+                                    await callback(self.MESSAGE, resp)
+                finally:
```
Comment (sfitts) [inline src/main/python/vantiqsdk.py:1460]: This is the new part; we pull the close handling up into the finally block.

File: src/main/python/vantiqsdk.py [hunk 3]
```diff
@@ -1478,7 +1479,8 @@ async def subscribe(self, path: str, params: dict, callback: Callable[[str, dict
         if target_namespace is not None:
             request_id += '@' + target_namespace
 
-        if request_id in self.subscriptions.keys():
+        # Check to see if we already have a subscription for this path
+        if self.subscriptions.get(request_id, False):
```
Comment (sfitts) [inline src/main/python/vantiqsdk.py:1483]: The previous code had a failure mode when the subscribe request failed with an exception and then was in a state that did not permit a retry of the subscription.

**Review by AlexBlumer — APPROVED** [review]
LGTM

**Review by mer2github — APPROVED** [review]
LGTM - where is the 12-hour test that validates the fix? :-)

**Comment by sfitts** [comment]
> LGTM - where is the 12-hour test that validates the fix? :-)

I thought about how to test, but I don't think the testing infrastructure lets us simulate an error from the network layer.  On my laptop I was able to simulate by simply breaking in the ping handler on the Java side and waiting until the python side timed out (in about 20s).  That's how I noticed the behavior difference between that case and what happens when you simply nuke the Java side (still closes, but no exception in that case for some reason).

**Pull request merged**